[18219] Documentation: add installation guide for Linux from sources (#61)
eProsima/Fast-DDS-QoS-Profiles-Manager: docs/rst/installation/linux_installation.rst

Linux installation from sources
===============================

> The *eProsima Fast DDS QoS Profiles Manager* tool suite is composed by many packages as described in the :ref:`home <qos_prof_description_>` section.

Author: Sphinx complaints about this, displaying a warning regarding the absence of a title:

`Failed to create a cross reference. A title or caption not found: qos_prof_description_`

I will let it as  `:ref:'Introduccion <qos_prof_description_>'`

        source fastdds_qosprofman_venv/bin/activate

The *Documentation* tool has several dependencies described in the ``requirements.txt`` project file.

.. literalinclude:: /requirements.txt
    :language: text

Install them using `pip3 <https://docs.python.org/3/installing/index.html>`_:

.. code-block:: bash

    pip3 install -r ~/fastdds_qosprofman_ws/src/Fast-DDS-QoS-Profiles-Manager/docs/requirements.txt

.. _installation_build_source_linux:

.. tabs::

    .. tab:: Building with colcon

        *eProsima Fast DDS QoS Profiles Manager* complete tool suite is build using `colcon <https://colcon.readthedocs.io/en/released/>`_  by running:

        .. code-block:: bash

            cd ~/fastdds_qosprofman_ws
            colcon build

        .. note::

            Being based on `CMake <https://cmake.org/>`_ it is possible to pass `CMake configuration options <https://colcon.readthedocs.io/en/released/reference/verb/build.html#cmake-specific-arguments>`_ to the :code:`colcon build` command.
            These options can be also passed through a `colcon.meta <https://colcon.readthedocs.io/en/released/user/configuration.html?highlight=colcon.meta#using-meta-files>`_ file.

>             For instance, the *Library* test suite building is enabled by building with the CMake option `EPROSIMA_BUILD_TESTS`, among adding the GTest dependency (``"googletest-distribution"``) in the *Library* `colcon.pkg <https://colcon.readthedocs.io/en/released/user/configuration.html?highlight=colcon.meta#colcon-pkg-files>`_ file. Also, colcon provides `certain arguments <https://colcon.readthedocs.io/en/released/reference/package-selection-arguments.html>`_ to perform package selection, excluding undesired packages or including specific packages in the build process.

Reviewer: The documentation test suite is enabled with the `EPROSIMA_BUILD_TEST` CMake option too. As the PR is being reviewed, I think we can already add the reference here.

>             For instance, the *Library* test suite building is enabled by building with the CMake option `EPROSIMA_BUILD_TESTS`, among adding the GTest dependency (``"googletest-distribution"``) in the *Library* `colcon.pkg <https://colcon.readthedocs.io/en/released/user/configuration.html?highlight=colcon.meta#colcon-pkg-files>`_ file. Also, colcon provides `certain arguments <https://colcon.readthedocs.io/en/released/reference/package-selection-arguments.html>`_ to perform package selection, excluding undesired packages or including specific packages in the build process.

Reviewer: Adding GTest to the colcon.pkg is only required if the GTest repo has been cloned into the workspace.

>             The following command would build the *Library* tool with its test suite.

Reviewer: ```suggestion
            For instance, the following command would build the *Library* tool with its test suite.
```

Author: I will change to `Thus, ...` because the previous paragraph starts also with `For instance, ...`

            .. code-block:: bash

                colcon build --packages-select fastdds_qos_profiles_manager_lib --cmake-args -DEPROSIMA_BUILD_TESTS=ON

        .. important::

>             The *Library* package is required for building the remain tool packages.

Reviewer: This should have been stated in the dependencies section. I am not against leaving also this notice.

    .. tab:: Building with CMake

        *eProsima Fast DDS QoS Profiles Manager* tool suite is build using `CMake <https://cmake.org/>`_ following these steps:

        **Library**

        .. code-block:: bash

            mkdir -p ~/fastdds_qosprofman_ws/build/lib
            cd ~/fastdds_qosprofman_ws/build/lib
            cmake ../../src/Fast-DDS-QoS-Profiles-Manager/lib -DCMAKE_INSTALL_PREFIX=../../install/lib
            cmake --build . --target install

        .. note::

>             If *Library* test suite compilation is needed, the enable test build flag `EPROSIMA_BUILD_TESTS` should be included while building the *Library* package:

Reviewer: ```suggestion
            If the *Library* test suite compilation is needed, the enable test build flag `EPROSIMA_BUILD_TESTS` should be included while configuring the building of the *Library* package:
```

            .. code-block:: bash

                cmake ../../src/Fast-DDS-QoS-Profiles-Manager/lib -DCMAKE_INSTALL_PREFIX=../../install/lib -DEPROSIMA_BUILD_TESTS=ON

        **CLI**

        .. code-block:: bash

            mkdir -p ~/fastdds_qosprofman_ws/build/cli
            cd ~/fastdds_qosprofman_ws/build/cli
            export CMAKE_PREFIX_PATH=${CMAKE_PREFIX_PATH}:~/fastdds_qosprofman_ws/install/lib
            cmake ../../src/Fast-DDS-QoS-Profiles-Manager/cli -DCMAKE_INSTALL_PREFIX=../../install/cli
            cmake --build . --target install

        .. note::

            Note that the already installed *Library* path has been sourced in the *CLI* build process with the command:

            .. code-block:: bash

                export CMAKE_PREFIX_PATH=${CMAKE_PREFIX_PATH}:~/fastdds_qosprofman_ws/install/lib

        **Documentation**

        .. important::

>             If the *Library* path has been sourced in the previous step, there is no need to include it again in the environment variable ``CMAKE_PREFIX_PATH``.
> 
>             If, on the other hand, it has not been sourced previously, as long as it is required also to build the *Documentation* package, it needs to be sourced.

Reviewer: ```suggestion
            If the *Library* path has been sourced in the previous step, there is no need to include it again in the environment variable ``CMAKE_PREFIX_PATH``.
            If, on the other hand, it has not been sourced previously, as it is also required to build the *Documentation* package, it needs to be sourced.
```

        .. code-block:: bash

            mkdir -p ~/fastdds_qosprofman_ws/build/docs
            cd ~/fastdds_qosprofman_ws/build/docs
            cmake ../../src/Fast-DDS-QoS-Profiles-Manager/docs -DCMAKE_INSTALL_PREFIX=../../install/docs
            cmake --build . --target install

Run CLI example
---------------

The ``fastddsqosprof`` executable file is generated in the installation path.
In order to run the CLI, please, ensure to source the environment:

.. tabs::

    .. tab:: Built with colcon

        .. code-block:: bash

            source ~/fastdds_qosprofman_ws/install/setup.bash

    .. tab:: Built with CMake

        .. code-block:: bash

            export LD_LIBRARY_PATH=~/fastdds_qosprofman_ws/install/lib/lib
            export PATH=$PATH:~/fastdds_qosprofman_ws/install/cli/bin

*CLI* should display its software version by running:

.. code-block:: bash

    fastddsqosprof -v

> 

Reviewer: Maybe a note here redirecting to the CLI example folder (even if there is only one right now) could be useful.

Local documentation
-------------------

The *Documentation* generated can be opened in a web browser locally by running:

.. tabs::

    .. tab:: Built with colcon

        .. code-block:: bash

            xdg-open ~/fastdds_qosprofman_ws/build/fastdds_qos_profiles_manager_docs/docs/html/index.html

    .. tab:: Built with CMake

        .. code-block:: bash

>             xdg-open ~/fastdds_qosprofman_ws/build/docs/docs/html/index.html

Reviewer: Maybe, instead of using the build folder, it is better to redirect to the install one.

    This tool is still in development process.

Documentation tool
^^^^^^^^^^^^^^^^^^

The *eProsima Fast DDS QoS Profiles Manager Documentation* depends on the *Library tool* and also on the following packages.

Doxygen
"""""""

`Doxygen <https://www.doxygen.nl/>`_ is a C++ documentation generator.
It is required to build the :ref:`Library API reference <fastdds_qos_profiles_manager_lib_api_reference>`.
In Ubuntu, the dependency can be installed running:

.. code-block:: bash

    sudo apt install doxygen

Python3 virtual environment (recommended)
"""""""""""""""""""""""""""""""""""""""""

It is recommended to install the *Documentation* tool dependencies in a `python3 <https://www.python.org/downloads/>`_ virtual environment.
That would avoid polluting the user's installation setup.

Deploy the project in a python3 virtual environment by running:

    .. code-block:: bash

        # Python3 virtual environment installation
        sudo apt install python3.10-venv

        # Virtual environment deployment
        cd <path_to_virtual_environment_directory>
        python3 -m venv fastdds_qosprofman_venv
        source fastdds_qosprofman_venv/bin/activate

    .. note::

>         As long as colcon performs a recursive search to build packages inside the workspace directory, it is recommended to deploy the virtual environment outside of the workspace directory in order to avoid conflicts.

Reviewer: Please, do not indent.

Author: 🤦🏻

>     .. code-block:: bash
> 
>         # Python3 virtual environment installation
>         sudo apt install python3.10-venv
> 
>         # Virtual environment deployment
>         cd <path_to_virtual_environment_directory>
>         python3 -m venv fastdds_qosprofman_venv
>         source fastdds_qosprofman_venv/bin/activate
> 
>     .. note::
> 
>         As long as colcon performs a recursive search to build packages inside the workspace directory, it is recommended to deploy the virtual environment outside of the workspace directory in order to avoid conflicts.

Reviewer: ```suggestion
    .. code-block:: bash

        # Python3 virtual environment installation
        sudo apt install python3.10-venv

        # Virtual environment deployment
        cd <path_to_virtual_environment_directory>
        python3 -m venv fastdds_qosprofman_venv
        source fastdds_qosprofman_venv/bin/activate

    .. note::

        As colcon performs a recursive search to build packages inside the workspace directory, it is recommended to deploy the virtual environment outside of the workspace directory in order to avoid conflicts.
```

            If, on the other hand, it has not been sourced previously, as long as it is required also to build the *Documentation* package, it needs to be sourced.

        .. code-block:: bash

            mkdir -p ~/fastdds_qosprofman_ws/build/docs
            cd ~/fastdds_qosprofman_ws/build/docs
            cmake ../../src/Fast-DDS-QoS-Profiles-Manager/docs -DCMAKE_INSTALL_PREFIX=../../install/docs
            cmake --build . --target install

Run CLI example
---------------

The ``fastddsqosprof`` executable file is generated in the installation path.
In order to run the CLI, please, ensure to source the environment:

.. tabs::

    .. tab:: Built with colcon

        .. code-block:: bash

            source ~/fastdds_qosprofman_ws/install/setup.bash

    .. tab:: Built with CMake

        .. code-block:: bash

            export LD_LIBRARY_PATH=~/fastdds_qosprofman_ws/install/lib/lib
            export PATH=$PATH:~/fastdds_qosprofman_ws/install/cli/bin

*CLI* should display its software version by running:

.. code-block:: bash

    fastddsqosprof -v

> Next steps: please look at the `examples <https://github.com/eProsima/Fast-DDS-QoS-Profiles-Manager/tree/main/cli/examples>`_. Each example should have its ``README`` explaining the use case and how to launch it.

Reviewer: ```suggestion
Next steps: please look at the `CLI examples <https://github.com/eProsima/Fast-DDS-QoS-Profiles-Manager/tree/main/cli/examples>`_.
```

        by running:

        .. code-block:: bash

            cd ~/Fast-DDS-QoS-Profiles-Manager
            colcon build --packages-select fastdds_qos_profiles_manager_lib fastdds_qos_profiles_manager_cli

        .. note::

            As happened in the *library* build process, if test compilation is needed, the enable test build flag
            `EPROSIMA_BUILD_TESTS` should be included too:

            .. code-block:: bash

                colcon build --packages-select fastdds_qos_profiles_manager_lib fastdds_qos_profiles_manager_cli --cmake-args -DEPROSIMA_BUILD_TESTS=ON

        The ``fastddsqosprof`` executable file is generated in the installation path.
        The following setup call must be run in order to set the *CLI* command:

        .. code-block:: bash

            source ~/Fast-DDS-QoS-Profiles-Manager/install/setup.bash

    .. tab:: Building with CMake

        *eProsima Fast DDS QoS Profiles Manager CLI* is build using `CMake <https://cmake.org/>`_ by running:

        .. code-block:: bash

            mkdir -p ~/Fast-DDS-QoS-Profiles-Manager/build/cli
            cd ~/Fast-DDS-QoS-Profiles-Manager/build/cli

>             CMAKE_PREFIX_PATH+=~/Fast-DDS-QoS-Profiles-Manager/install/lib

Reviewer: ```suggestion
            CMAKE_PREFIX_PATH+=:~/Fast-DDS-QoS-Profiles-Manager/install/lib
```

This instruction only works if the environment variable has been exported (declared) previously. I suggest using:

```suggestion
            export CMAKE_PREFIX_PATH=${CMAKE_PREFIX_PATH}:~/Fast-DDS-QoS-Profiles-Manager/install/lib
```

This way the variable is ensured to be declared.

            cmake ../../src/Fast-DDS-QoS-Profiles-Manager/cli -DCMAKE_INSTALL_PREFIX=../../install/cli
            cmake --build . --target install

        .. note::

            As happened in the *library* build process, if test compilation is needed, the enable test build flag
            `EPROSIMA_BUILD_TESTS` should be included too:

            .. code-block:: bash

>                 cmake ../../src/Fast-DDS-QoS-Profiles-Manager/cli -DCMAKE_INSTALL_PREFIX=../../install/cli -DEPROSIMA_BUILD_TESTS=ON

Reviewer: There are no test suite for the CLI for the moment. Please remove until there is.

        The ``fastddsqosprof`` executable file is generated in the installation path.

>         The following setup call must be run in order to set the *CLI* command:

Reviewer: ```suggestion
        In order to run the CLI, please, ensure to source the environment:
```

    .. code-block:: bash

        sudo apt install python3
        cd ~/Fast-DDS-QoS-Profiles-Manager
        python3 -m venv fastdds_qos_profiles_manager_venv
        source fastdds_qos_profiles_manager_venv/bin/activate

Install the dependencies using `pip3 <https://docs.python.org/3/installing/index.html>`_:

.. code-block:: bash

    cd ~/Fast-DDS-QoS-Profiles-Manager/src/Fast-DDS-QoS-Profiles-Manager/docs
    pip3 install -r requirements.txt

.. _doc_building_sl:

Build
^^^^^

.. tabs::

    .. tab:: Building with colcon

        *eProsima Fast DDS QoS Profiles Manager documentation* is build using
        `colcon <https://colcon.readthedocs.io/en/released/>`_ by running:

        .. code-block:: bash

            cd ~/Fast-DDS-QoS-Profiles-Manager
            colcon build --packages-select fastdds_qos_profiles_manager_lib fastdds_qos_profiles_manager_docs

    .. tab:: Building with CMake

        *eProsima Fast DDS QoS Profiles Manager documentation* is build using `CMake <https://cmake.org/>`_ by running:

        .. code-block:: bash

            mkdir -p ~/Fast-DDS-QoS-Profiles-Manager/build/docs
            cd ~/Fast-DDS-QoS-Profiles-Manager/build/docs

>             CMAKE_PREFIX_PATH+=~/Fast-DDS-QoS-Profiles-Manager/install/lib

Reviewer: Please, apply the same suggestion. Although, as it is being run in the same snippet, this step should be done only once, but I suggest adding a note explaining that this is required so the library dependency is found.

    Please check the :ref:`colcon build procedure <installation_build_source_linux>` for further information.

CLI tool
^^^^^^^^

The *eProsima Fast DDS QoS Profiles Manager CLI* depends on the tool suite *Library* and also on the following package.

Docopt
""""""

`docopt <http://docopt.org/>`_ is a CLI description language that defines the interface of the command line applications and generates a parser for it automatically.
In Ubuntu, the dependency can be installed running:

.. code-block:: bash

    sudo apt install libdocopt-dev

GUI tool
^^^^^^^^

.. warning::

    This tool is still in development process.

Documentation tool
^^^^^^^^^^^^^^^^^^

The *eProsima Fast DDS QoS Profiles Manager Documentation* depends on the tool suite *Library* and also on the following package.

Doxygen
"""""""

`Doxygen <https://www.doxygen.nl/>`_ is a C++ documentation generator.
It is required to build the :ref:`Library API reference <fastdds_qos_profiles_manager_lib_api_reference>`.
In Ubuntu, the dependency can be installed running:

.. code-block:: bash

    sudo apt install doxygen

> Python3 virtual environment (recomended)

Reviewer: I do not think this is the correct title. These dependencies are required. What is it optional is either installed them in a virtual environment or in the host directly. I suggest renaming it to `Python3 package dependencies`.

The `python3-venv` is recommended. Maybe this section can be kept but it needs to be split from the other mandatory dependencies.

""""""""""""""""""""""""""""""""""""""""

It is recommended to install the *Documentation* tool dependencies in a `python3 <https://www.python.org/downloads/>`_ virtual environment.
That would avoid polluting the user's installation setup.

> Also, the *Documentation* tool has several dependencies described in the ``requirements.txt`` project file.

Reviewer: I think this should be reordered. The last sentence must be the first one and then explain the options to install them.

.. literalinclude:: /requirements.txt
    :language: text

.. important::

>     Once :ref:`the repository is downloaded <installation_download_project_source_linux>`, deploy the project in a python3 virtual environment and install the ``requirements.txt`` dependencies by running:

Reviewer: ```suggestion
    Once :ref:`the repository is downloaded <installation_download_project_source_linux>`, the ``requirements.txt`` dependencies file can be installed by running:
```

    .. code-block:: bash

>         # Python3 installation

Reviewer: ```suggestion
        # Python3 virtual environment installation
```

This must be in the other dependency section.

        sudo apt install python3.10-venv

        # Virtual environment deployment
        cd <path_to_virtual_environment_directory>
        python3 -m venv fastdds_qosprofman_venv
        source fastdds_qosprofman_venv/bin/activate

        # Required dependencies installation
        pip3 install -r ~/fastdds_qosprofman_ws/src/Fast-DDS-QoS-Profiles-Manager/docs/requirements.txt

    .. note::

>         The ``<path_to_virtual_environment_directory>`` could be inside the workspace directory if project is built using CMake.

Reviewer: This is too much information. I would only leave the notice about colcon and the virtual environment. This note I would think that it makes more sense in the section of installing the virtual environment.

        As long as colcon performs a recursive search to build packages inside the workspace directory, it is recommended to deploy the virtual environment outside of the workspace directory in order to avoid conflicts.

            .. tabs::

                .. tab:: Building with colcon

                        .. code-block:: bash

                            # Virtual environment deployment
                            cd ~

                .. tab:: Building with CMake

                        .. code-block:: bash

                            # Virtual environment deployment
                            cd ~/fastdds_qosprofman_ws

> 

Reviewer: Please remove

------------------------

Create a workspace directory and download the project:

.. code-block:: bash

    mkdir -p ~/fastdds_qosprofman_ws/src
    cd ~/fastdds_qosprofman_ws/src
    git clone https://github.com/eProsima/Fast-DDS-QoS-Profiles-Manager.git

.. _installation_build_source_linux:

.. tabs::

    .. tab:: Building with colcon

        *eProsima Fast DDS QoS Profiles Manager* complete tool suite is build using `colcon <https://colcon.readthedocs.io/en/released/>`_  by running:

        .. code-block:: bash

            cd ~/fastdds_qosprofman_ws
            colcon build

        .. note::

            Being based on `CMake <https://cmake.org/>`_ it is possible to pass `CMake configuration options <https://colcon.readthedocs.io/en/released/reference/verb/build.html#cmake-specific-arguments>`_ to the :code:`colcon build` command.
            These options can be also passed through a `colcon.meta <https://colcon.readthedocs.io/en/released/user/configuration.html?highlight=colcon.meta#using-meta-files>`_ file.

            For instance, the *Library* and *Documentation* test suites building is enabled by building with the CMake option `EPROSIMA_BUILD_TESTS`.

>             If GTest repository has been included in the workspace directory, add the GTest dependency (``"googletest-distribution"``) in both *Library* and *Documentation* `colcon.pkg <https://colcon.readthedocs.io/en/released/user/configuration.html?highlight=colcon.meta#colcon-pkg-files>`_ files.

Reviewer: As the documentation depends on the library, it is enough adding it only to the library package. I think we should keep this as simple as possible.

            cmake --build . --target install

Run CLI example
---------------

The ``fastddsqosprof`` executable file is generated in the installation path.
In order to run the CLI, please, ensure to source the environment:

.. tabs::

    .. tab:: Built with colcon

        .. code-block:: bash

            source ~/fastdds_qosprofman_ws/install/setup.bash

    .. tab:: Built with CMake

        .. code-block:: bash

            export LD_LIBRARY_PATH=~/fastdds_qosprofman_ws/install/lib/lib
            export PATH=$PATH:~/fastdds_qosprofman_ws/install/cli/bin

*CLI* should display its software version by running:

.. code-block:: bash

    fastddsqosprof -v

.. note::

    A complete example of the *CLI* is included in the *eProsima Fast DDS QoS Profiles Manager* repository.
    It depends on `jq <https://stedolan.github.io/jq/>`_ in order to parse the deployment information contained in a JSON file.
    The instructions below would run the example and generate several XML configuration files.

    .. code-block:: bash

        sudo apt install jq
        cd ~/fastdds_qosprofman_ws/src/Fast-DDS-QoS-Profiles-Manager/cli/examples/ExternalLocatorFeatureConfigurationExample

>         bash cli_example_script.sh

Reviewer: Keep it simple: just redirect to the examples folder. More examples can be included later on and this will become obsolete. The idea is similar to [this section](https://docs.ros.org/en/foxy/Installation/Alternatives/Ubuntu-Development-Setup.html#next-steps-after-installing). Next steps: please look at the examples. Each example should have its README explaining the use case and how to launch it.

> 

Reviewer: Please, remove extra blank lines
